### DISCUSSION TITLE
mitmproxy not redirecting to a captive portal page

### DISCUSSION THREAD

**Discussion by changsin** [mitmproxy/mitmproxy#7035] [Q&A] [answered]
When I try to use mitmproxy on a public network that enforces a captive portal page (for a login screen or terms and services page), any URL does not get redirected to the captive portal page. Instead, I see a blank page from the browser.
The workaround that I found is to manually go to a site like http://httpforever.com/. This will trigger the browser to redirect to the captive portal page. 

Here are my questions:
1. Why is the captive portal page not opening?
2. Why does the workaround work?
3. Is there a better solution?

Here is my current mitmproxy version, but should be the same for other versions:

Mitmproxy: 10.3.1
Python:    3.10.11
OpenSSL:   OpenSSL 3.2.2 4 Jun 2024
Platform:  macOS-14.5-arm64-arm-64bit

**Comment by Prinzhorn** [comment]
So you're saying _without_ mitmproxy you get the captive portal page even for HTTPS websites?

I don't know how these pages are implemented, but does `--set connection_strategy=lazy` help? How did you configure the proxy?

https://www.chromium.org/chromium-os/chromiumos-design-docs/network-portal-detection/

> To the extent that connections using a web proxy are never marked as portal but instead immediately marked as online this inhibits shill's ability to choose the best connection/service.

**Reply by changsin** [reply]
Yes, without mitmproxy, the captive portal page displays (a pop-up window). With mitmproxy, there is no pop-up dialog and HTTPS access fails silently. Only when I access HTTP site, it directs to the captive portal page.
I tried connection_strategy=lazy, but it did not make any difference.

**Reply by Prinzhorn** [reply]
If you are able to test that: does it work with _any_ other proxy (just a regular one)?

**Reply by changsin** [reply]
What would be a good proxy to try? The OS causes the pop-up window because I see a call was made to http://www.msftconnecttest.com/ when I switch to a public network with a captive portal. I could see this from mitmproxy log but when mitmproxy is on, the pop-up window is somehow suppressed and not shown (on macOS).

**Reply by Prinzhorn** [reply]
You could setup Squid or any other proxy on a throwaway VPS that you destroy after you've tested it. E.g. https://www.digitalocean.com/community/tutorial-collections/how-to-set-up-squid-proxy-for-private-connections

If it behaves the same, then this is a general issue with Edge+Proxy+captive portal page and not related to mitmproxy.

**Reply by changsin** [reply, marked as answer]
I found a workaround. In your add-on request method, You can check if an HTTP GET call was made to one of the known captive portal sites and, if it is, return immediately.
```
    def request(self, flow: http.HTTPFlow):
        if flow.request.method in ['GET']:
            if flow.request.scheme == 'http' and 'www.msftconnecttest.com' in flow.request.url:
                    return
```
